akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala

```diff
@@ -318,13 +348,25 @@ private[stream] object ConnectionSourceStage {
           if (isClosed(bytesOut)) connection ! ResumeReading
           else push(bytesOut, data)
 
+        case WriteDelay =>
+          delayCount += 1
+          writeDelayCountDown -= 1
+          if (writeDelayCountDown == 0 || previousWriteBufferSize == writeBuffer.size || writeBuffer.size >= writeBufferSize) {
+            sendWriteBuffer()
+          } else {
+            previousWriteBufferSize = writeBuffer.size
+            connection ! writeDelayMessage
```

Comment:
Ok, so the logic is that we loop for a limited number of times using the mailbox as long as we have collected some more data in the last round.

That was my original idea in the coalesce writes patch as well but only with a single loop (which already works quite well because we will allow the stream interpreter to run at least on all already buffered data if the number of steps is < syncProcessingLimit).

Have you checked how much the extra loops after the first one help? I could imagine that this number together with actor throughput and streams syncProcessingLimit throughput are somewhat hard to tweak.

Reply:
> Ok, so the logic is that we loop for a limited number of times using the mailbox as long as we have collected some more data in the last round.

That is right.

> Have you checked how much the extra loops after the first one help?

There was already batching collecting additional elements during the "first" write that worked pretty well. This additional delay after the ack seems to double the average bytes/write both when running locally and in staging. I'm not sure I tested the impact of the 20 count down or if it never reached more than 1 extra roundtrip.